### PR TITLE
DDPB-3670: Move and reimplement one audit log from client to API

### PR DESCRIPTION
**Pull request by acsauk-mojd**
## Purpose
As part of Symfony upgrade preparations, we've decided to move the audit logging from the client-side to the API side in order to move away from overwriting services in the DI container during tests. This approach also allows us to write the business logic (and tests) for certain entity update events once and then forget about it rather than having to manually implement it in each place in the app that updates the entity.

Fixes DDPB-3670

## Approach
I've hooked into the Doctrine lifecycle events rather than adding this to an entity class or as a service. This means that any time a user entity email address is updated in any part of the app we will have a log created.

The following subtask under the main logging and emails ticket will move the remainder of the log events over.

## Learning
I'd forgotten that when dealing with the services.yaml and config files a cache reset is required to recreate the new service container in order to reflect any changes.

## Checklist
- [x] I have performed a self-review of my own code
- [ ] I have updated documentation (Confluence/ADR/tech debt doc) where relevant
- [x] I have added tests to prove my work
- [ ] The product team have approved these changes

### CODE REVIEW

**Review by jamesrwarren — COMMENTED** [review]
Looks good. Will let the devs have a look over it though before approving

**Review by Gugandeep — COMMENTED** [review]
Looks good!

I see you've implemented wait for it which is great. Gives us a head start on the following ticket: [DDPB-3518](https://opgtransform.atlassian.net/browse/DDPB-3518)

File: api/docker/confd/templates/parameters.yml.tmpl [hunk 1]
```diff
@@ -3,6 +3,18 @@
 
 # This file is a "template" of what your parameters.yml file should look like
 parameters:
+    cloudwatch_logs_client_params:
+       version: 'latest'
+       region: 'eu-west-1'
+{{ if exists "/cloudwatch/logs/localstack" }}
```
Comment (coderste) [inline api/docker/confd/templates/parameters.yml.tmpl:9]: Could you explain this part for me please? When would this exist and when wouldn't it?

File: api/src/AppBundle/Service/Audit/AuditEvents.php [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            'changed_on' => $this->dateTimeProvider->getDateTime()->format(DateTime::ATOM),
+            'changed_by' => $changedBy,
+            'subject_full_name' => $subjectFullName,
+            'subject_role' => $subjectRole,
+        ];
+
+        return $event + $this->baseEvent(AuditEvents::EVENT_USER_EMAIL_CHANGED);
+    }
+
+    public function clientEmailChanged(
```
Comment (coderste) [inline api/src/AppBundle/Service/Audit/AuditEvents.php:84]: No PHPDoc above this function

Comment (coderste) [reply api/src/AppBundle/Service/Audit/AuditEvents.php:84]: Also do we know the type this function returns? If so should we typehint it?

File: api/src/AppBundle/Service/Audit/AwsAuditLogHandler.php [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    public function __construct(CloudWatchLogsClient $client, $group, $level = Logger::NOTICE, $bubble = true)
+    {
+        $this->client = $client;
+        $this->group = $group;
+
+        parent::__construct($level, $bubble);
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (coderste) [inline api/src/AppBundle/Service/Audit/AwsAuditLogHandler.php:44]: Like this :D didn't even know it was a thing 👍🏻

**Review by coderste — CHANGES_REQUESTED** [review]
Just the PHPDocs missing on the one function other than that this is great work 👏🏻 👍🏻

File: api/docker/confd/templates/parameters.yml.tmpl (hunk 1 above)
Comment (acsauk) [reply api/docker/confd/templates/parameters.yml.tmpl:9]: So we use confd to handle some of the config values here in different environments. I'm not a huge fan of it as the syntax isn't very clear with what its doing and Symfony has its own environment system that we can hook into with things like `env.test` and `config.test` etc but that's a different discussion 😂 

This is basically saying if there's an environment variable called CLOUDWATCH_LOGS_LOCALSTACK then compile the values down. We only have this variable in environments where we want localstack to run so its excluded any places that use real infra. The confd process itself runs after [composer is run](https://github.com/ministryofjustice/opg-digideps/blob/main/client/composer.json#L38) as a post hook and is something we'll need to update when we do the Symfony upgrade as the package that handles that won't work in Symfony 4+ seeing as they use [Symfony Flex](https://symfony.com/doc/current/setup/flex.html) instead.

File: api/src/AppBundle/Service/Audit/AuditEvents.php (hunk 2 above)
Comment (acsauk) [reply api/src/AppBundle/Service/Audit/AuditEvents.php:103]: Good spot - actioned 👍

File: api/src/AppBundle/Service/Audit/AwsAuditLogHandler.php (hunk 3 above)
Comment (acsauk) [reply api/src/AppBundle/Service/Audit/AwsAuditLogHandler.php:44]: Yeah - one thing in PHP that encourages us to type less 🙃

File: api/docker/confd/templates/parameters.yml.tmpl (hunk 1 above)
Comment (coderste) [reply api/docker/confd/templates/parameters.yml.tmpl:9]: Ah! okay that makes more sense to me now 👍🏻

**Review by coderste — APPROVED** [review]
LGTM 👍🏻

**Review by Gugandeep — COMMENTED** [review]
Latest changes looks good!

**Pull request merged**